### PR TITLE
split service locator and cluster locators

### PR DESCRIPTION
**Pull request by mclarke47**
Break the coupling between how we decide whether a service is on a cluster and how we load cluster configuration into the  kubernetes backend.

This should hopefully make it easier to add implementations for cluster locators and service locators.

## config migration

`clusterLocatorMethod` is now 2 fields

`serviceLocatorMethod` determines how we figure out what cluster a service is on

We can only have 1 `serviceLocatorMethod` as we need an authoritative source of which cluster we should look on.

`clusterLocatorMethods` determines how we load cluster configurations so that we can create kubernetes clients.

We can have multiple `clusterLocatorMethods` as we can load cluster information from the config file or a provider's (GCP, AWS etc..) API, or both.

### Before this PR

```yaml
kubernetes:
  clusterLocatorMethod: 'configMultiTenant'
  clusters:
    - url: https://199.168.64.2:8443
      name: minikube
      serviceAccountToken: 'my-token'
      authProvider: 'serviceAccount'
```

### After this PR

```yaml

kubernetes:
  serviceLocatorMethod: 'multiTenant'
  clusterLocatorMethods:
    - 'config'
  clusters:
    - url: https://199.168.64.2:8443
      name: minikube
      serviceAccountToken: 'my-token'
      authProvider: 'serviceAccount'
```
## Testing
I loaded up the dice roller, connected to GKE and minikube and the UI loaded fine

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/kubernetes-backend/README.md [hunk 1]
```diff
@@ -8,21 +8,33 @@ It responds to Kubernetes requests from the frontend.
...
+
+This clusterLocatorMethod will read cluster information in from config
+
 Example:
 
 ```yaml
 kubernetes:
-  clusterLocatorMethod: 'configMultiTenant'
+  serviceLocatorMethod: 'multiTenant'
+  clusterLocatorMethods:
```
Comment (Rugvip) [inline plugins/kubernetes-backend/README.md:36]: What do you think about merging this into the config below instead? Making it more of a `clusterSources`-type thing, each object having its own `type` or `method`. Something like

```yaml
clusters:
  - type: config
    options:
      url: http://127.0.0.1:9999
      name: minikube
      serviceAccountToken: <TOKEN FROM STEP 4>
      authProvider: 'serviceAccount'
  - type: config
    options:
      url: http://127.0.0.2:9999
      name: gke-cluster-1
      authProvider: 'google'
  - type: kubeconfig
    options:
      file: ~/.kube/config
  - type: http # if we want to attempt to standardize this
    options:
      url: http://k8s-discovery-service.internal.example.com/v1/cluster-discovery
```

File: plugins/kubernetes-backend/src/types/types.ts [hunk 2]
```diff
@@ -118,8 +118,13 @@ export interface KubernetesFetcher {
 }
 
 // Used to locate which cluster(s) a service is running on
-export interface KubernetesClusterLocator {
-  getClusterByServiceId(serviceId: string): Promise<ClusterDetails[]>;
+export interface KubernetesServiceLocator {
+  getClustersByServiceId(serviceId: string): Promise<ClusterDetails[]>;
```
Comment (Rugvip) [inline plugins/kubernetes-backend/src/types/types.ts:122]: Still not super thrilled about the `serviceId`. Thoughts on simply passing an `Entity` here instead?

**Review by Rugvip — APPROVED** [review]
Awesome stuff 😁

File: plugins/kubernetes-backend/README.md (hunk 1 above)
Comment (mclarke47) [reply plugins/kubernetes-backend/README.md:36]: Yes, I think we probably should go in this direction, but I'm going to wait until I try to add the gcloud implementation before coming up with the config values

File: plugins/kubernetes-backend/src/types/types.ts (hunk 2 above)
Comment (mclarke47) [reply plugins/kubernetes-backend/src/types/types.ts:122]: I think this makes sense, we are doing a POST now we should post the entity with the auth details.

**Comment by mclarke47** [comment]
Gonna merge this and come back for the suggestions on my next hackday, next wednesday

File: plugins/kubernetes-backend/src/types/types.ts (hunk 2 above)
Comment (Rugvip) [reply plugins/kubernetes-backend/src/types/types.ts:122]: Yep, although we can also have the layer outside this go fetch the entity data from the catalog by name too.

```ts
router.get('/mah-resources/:entityName', (req, res) => {
  const entity = await catalogClient.getEntityByName(req.params.entityName);
  const clusters = await serviceLocator.getClustersByEntity(entity);
  ...
})
```

**Pull request merged**